realtek: improve SoC detection, read revision

janh opened the pull request:
Instead of the hard-coded model numbers, read the full model number from the register. Also remove detection of some unsupported SoCs.

Read and print the revision, which can be useful sometimes when there are relevant differences (see for example https://github.com/openwrt/openwrt/commit/d4bf16a9e1618347709b0dccae71c273e6291a61).

Tested on:
- HPE 1920-8G: `SoC Type: Realtek RTL8380M rev C`
- HPE 1920-16G: `SoC Type: Realtek RTL8382M rev C`
- HPE 1920-48G: `SoC Type: Realtek RTL8393M rev A`
- XGS1010-12: `SoC Type: Realtek RTL9302B rev A`

janh committed:
realtek: simplify SoC detection

Read model name from the register instead of using hard-coded values.

Also remove detection of the unsupported Realtek ESW/SSW SoCs. The Fast
Ethernet variants of the Maple and Cypress series stay for now, but are
moved to the RTL8380/RTL8390 families.

Signed-off-by: Jan Hoffmann <jan@3e8.eu>
Link: https://github.com/openwrt/openwrt/pull/19653
Signed-off-by: Robert Marko <robimarko@gmail.com>

janh committed:
realtek: extend SoC information

Add SoC revision, CPU part number, and a flag for engineering samples to
the rtl83xx_soc_info structure.

Also extend the system type string to include this information.

Signed-off-by: Jan Hoffmann <jan@3e8.eu>
Link: https://github.com/openwrt/openwrt/pull/19653
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul commented:
Thanks. Wanted to do this a long time but never had the usecase for it. Only had time to code a simple version in the rt-loader. Can you add CHIP_VER too so we get something like "6275C". See https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=ccbff8bbdd1aaca3ad76fc129b8d53effb161183

plappermaul reviewed (approved):
Independent from a to-be-added chip version ... approved.

janh committed:
realtek: move and clean up CHIP_INFO register definitions

Move the definitions to mach-rtl83xx.h, so they can be used during init
to read more detailed SoC information. Also rename the RTL931X register,
as it has the same address on all RTL93xx.

Signed-off-by: Jan Hoffmann <jan@3e8.eu>
Link: https://github.com/openwrt/openwrt/pull/19653
Signed-off-by: Robert Marko <robimarko@gmail.com>

janh commented:
HPE-1920-8G: `SoC Type: Realtek RTL8380M rev C (6275)`
HPE-1920-16G: `SoC Type: Realtek RTL8382M rev C (6275)`
HPE-1920-48G: `SoC Type: Realtek RTL8393M rev A (6290)`
XGS1010-12: `SoC Type: Realtek RTL9302B rev A (6487)`

plappermaul commented:
Thank you.

plappermaul commented on target/linux/realtek/files-6.12/arch/mips/rtl838x/prom.c:
```diff
@@ -96,125 +99,74 @@ void __init device_tree_init(void)
+			 soc_info.id, suffix);
+	} else {
+		snprintf(soc_name, sizeof(soc_name), "RTL%04X", soc_info.id);
+	}
```
suffix = val ? val + 'A' - 1 : 0; ist not 100% equivalent but would allow for only one snprintf and easy readability.

hailfinger replied:
> Thank you! I didn't realize that printf would just omit a 0 char value entirely.

The 0 char value is not omitted, but the string gets null-terminated one char earlier by the 0 char value which is OK in this specific case: the 0 char is the last char in the string and the snprintf return value is not checked.

robimarko commented:
Thanks! Rebased on top of main and merged!
